refactor: show planned path in enum class changes for maintainer review

dksmiffs opened the pull request:
Work started on #22, this draft intended for early maintainer feedback on approach.

dksmiffs committed:
refactor: show planned path in enum class changes for maintainer review

carlocorradini commented:
@dksmiffs I would also add the type to the enum class if mapping to a known DIS type (and therefore knowing the maximum number of elements)

dksmiffs committed:
refactor: show more enum class progress, need feedback on FactoryDecode call

dksmiffs commented:
@carlocorradini - in 6fb761f, please note line 99 in `src/DataTypes/StandardVariable.cpp`. I'd like your feedback on what I should do there. `StandardVariableType` is a `KUINT32` enum, but FactoryDecode takes a `KINT32`. I obviously don't like that code the way I have it written.

dksmiffs committed:
refactor: propose a second template argument for FactoryDecoder...

This change is needed to allow
StandardVariable::FactoryDecodeStandardVariable() to pass a KUINT32
as the first parameter to FactoryDecode, since the underlying type of
the StandardVariableType enum class is KUINT32.

karljj1 reviewed (approved):
Very nice. Thanks!

carlocorradini commented:
> @carlocorradini - in [6fb761f](https://github.com/karljj1/kdis/commit/6fb761f30a60d76b45a7aa22205ff029d7e910a8), please note line 99 in `src/DataTypes/StandardVariable.cpp`. ~I'd like your feedback on what I should do there.~ `StandardVariableType` is a `KUINT32` enum, but FactoryDecode takes a `KINT32`. ~I obviously don't like that code the way I have it written.~
> 
> Update: see [197acba](https://github.com/karljj1/kdis/commit/197acbacf83b52d6458ad826015f0bba34a7ba2f) below for my proposed fix to this issue.

~Because the generic represents the underlying type of an enum, why not set `KUINT32` as the default value? What do you think?~
Furthermore, what do you think if we take directly the enum type (rather than the primitive underlying type) as generic and use a combination of [std::is_enum](https://en.cppreference.com/w/cpp/types/is_enum) (static assertion that the generic is an enum) and [std::underlying_type](https://en.cppreference.com/w/cpp/types/underlying_type)?

dksmiffs commented:
Sounds good, @carlocorradini.
